sycl/include/sycl/ext/oneapi/experimental/user_defined_reductions.hpp

```diff
@@ -0,0 +1,116 @@
+
+#pragma once
+
+#include <sycl/detail/defines.hpp>
+#include <sycl/group_algorithm.hpp>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace ext::oneapi::experimental {
+
+// ---- reduce_over_group
+template <typename GroupHelper, typename T, typename BinaryOperation>
+sycl::detail::enable_if_t<(is_group_helper_v<GroupHelper>), T>
+reduce_over_group(GroupHelper group_helper, T x, BinaryOperation binary_op) {
+  if constexpr (sycl::detail::is_native_op<T, BinaryOperation>::value) {
+    return sycl::reduce_over_group(group_helper.get_group(), x, binary_op);
+  }
+#ifdef __SYCL_DEVICE_ONLY__
+  T *Memory = reinterpret_cast<T *>(group_helper.get_memory().data());
+  auto g = group_helper.get_group();
+  Memory[g.get_local_linear_id()] = x;
+  group_barrier(g);
+  T result = Memory[0];
+  if (g.leader()) {
+    for (int i = 1; i < g.get_local_linear_range(); i++) {
+      result = binary_op(result, Memory[i]);
+    }
+  }
+  group_barrier(g);
+  return group_broadcast(g, result);
+#else
+  std::ignore = group_helper;
+  throw runtime_error("Group algorithms are not supported on host.",
+                      PI_ERROR_INVALID_DEVICE);
+#endif
+}
+
+template <typename GroupHelper, typename V, typename T,
+          typename BinaryOperation>
+sycl::detail::enable_if_t<(is_group_helper_v<GroupHelper>), T>
+reduce_over_group(GroupHelper group_helper, V x, T init,
+                  BinaryOperation binary_op) {
+  if constexpr (sycl::detail::is_native_op<V, BinaryOperation>::value &&
+                sycl::detail::is_native_op<T, BinaryOperation>::value) {
+    return sycl::reduce_over_group(group_helper.get_group(), x, init,
+                                   binary_op);
+  }
+#ifdef __SYCL_DEVICE_ONLY__
+  return binary_op(init, reduce_over_group(group_helper, x, binary_op));
+#else
+  std::ignore = group_helper;
+  throw runtime_error("Group algorithms are not supported on host.",
+                      PI_ERROR_INVALID_DEVICE);
+#endif
+}
+
+// ---- joint_reduce
+template <typename GroupHelper, typename Ptr, typename BinaryOperation>
+sycl::detail::enable_if_t<(is_group_helper_v<GroupHelper> &&
+                           sycl::detail::is_pointer<Ptr>::value),
+                          typename sycl::detail::remove_pointer<Ptr>::type>
+joint_reduce(GroupHelper group_helper, Ptr first, Ptr last,
+             BinaryOperation binary_op) {
+  if constexpr (sycl::detail::is_native_op<
+                    typename sycl::detail::remove_pointer<Ptr>::type,
+                    BinaryOperation>::value) {
+    return sycl::joint_reduce(group_helper.get_group(), first, last, binary_op);
+  }
+#ifdef __SYCL_DEVICE_ONLY__
+  using T = typename std::remove_pointer<Ptr>::type;
+  auto g = group_helper.get_group();
+  T partial = *(first + g.get_local_linear_id());
+  Ptr second = first + g.get_local_linear_range();
+  sycl::detail::for_each(g, second, last,
+                         [&](const T &x) { partial = binary_op(partial, x); });
```

Comment:
> (@aelovikov-intel) This seems to be linear and not logarithmic in complexity. I'd expect something like https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/reduction.hpp#L1810-L1818 might be applicable here, although I'm not sure what gives better performance for different work sizes.
>
> (@AlexeySachkov) I would suggest to put a comment here for now. We don't yet have any data from actual uses of this new function about which use cases (data type, wg sizes) are the most popular ones.

This implementation is based on implementation of `sycl::joint_reduce`, AFAIK no customer's request to improve its performance. But will add a comment.

Reply:
Comment added in https://github.com/intel/llvm/pull/7587/commits/6af7a5ef6efa44637783fb25a1fced629129b78e